### PR TITLE
Redisearch index in operator

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR allows index search optimization to transform `IN` operator into a multiple `OR` sequence, in case of applicable array and index.
The query `MATCH (p:Person) where p.age in range(0,2)` will be transforming its filter tree to a `((p.age=0 OR  p.age=1) OR p.age=2)`, in the case that `:Person(Age)` is indexed.

### CODE REVIEW

File: src/filter_tree/filter_tree.c [hunk 1]
```diff
@@ -273,8 +273,8 @@ void _FilterTree_CollectAttributes(const FT_FilterNode *root, rax *attributes) {
 		AR_EXP_CollectAttributes(root->pred.rhs, attributes);
 		break;
 	}
-	default: {
-		assert(0);
+	case FT_N_EXP: {
+		AR_EXP_CollectAttributes(root->exp.exp, attributes);
 		break;
 	}
```
Comment (swilly22) [inline src/filter_tree/filter_tree.c]: Please keep the default in addition to `case FT_N_EXP`.

File: src/execution_plan/optimizations/utilize_indices.c [hunk 2]
```diff
@@ -28,12 +29,25 @@ int _reverseOp(int op) {
...
+		if(filter->pred.rhs->operand.type == AR_EXP_VARIADIC) {
+			// Swap.
+			AR_ExpNode *tmp = filter->pred.rhs;
+			filter->pred.rhs = filter->pred.lhs;
+			filter->pred.lhs = tmp;
+			filter->pred.op = _reverseOp(filter->pred.op);
+		}
+		break;
+	}
+	case FT_N_COND: {
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: I'm not sure if it is possible for `filter` to be of type `FT_N_COND` at this point,
have you seen this case being hit?

File: src/execution_plan/optimizations/utilize_indices.c [hunk 3]
```diff
@@ -206,16 +269,25 @@ bool _applicableFilter(Index *idx, OpFilter *filter) {
 	// Make sure the filter root is not a function.
 	// TODO: As a result of issue 667, transform "IN" into a sequence of "OR" to use in RedisSearch.
 	if(filter_tree->t == FT_N_EXP) {
-		res = false;
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: As we only allow `FT_N_EXP` to represent an `in` expression, move `if(!strcasecmp(exp->op.func_name, "in") == 0) return false;` from `_validateInExpression` here.

File: src/execution_plan/optimizations/utilize_indices.c [hunk 4]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
 	return true;
 }
 
+static bool _validateInExpression(AR_ExpNode *exp) {
+	if(!strcasecmp(exp->op.func_name, "in") == 0) return false;
+	assert(exp->op.child_count == 2);
+	AR_ExpNode *list = exp->op.children[1];
+	if(list->type != AR_EXP_OPERAND || list->operand.type != AR_EXP_CONSTANT) return false;
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: `if(list->operand.type != AR_EXP_CONSTANT) return false;` ?
why check for `AR_EXP_OPERAND` ?

File: src/execution_plan/optimizations/utilize_indices.c [hunk 5]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
...
+	assert(list->operand.constant.type == T_ARRAY);
+	SIValue listValue = list->operand.constant;
+	if(SIArray_Length(listValue) == 0) return false;
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: Empty list should be converted into an empty RediSearch iterator

File: src/execution_plan/optimizations/utilize_indices.c [hunk 6]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
...
+	uint listLen = SIArray_Length(listValue);
+	for(uint i = 0; i < listLen; i++) {
+		// Clone value since the array might be free later.
+		SIValue v = SI_CloneValue(SIArray_Get(listValue, i));
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: Do not clone at this point, we only care for validation at this point.

File: src/execution_plan/optimizations/utilize_indices.c [hunk 7]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
...
+		memcpy(lhs, inOp->op.children[0], sizeof(AR_ExpNode));
+		AppendRightChild(orNode, FilterTree_CreatePredicateFilter(OP_EQUAL, lhs, constant));
+		root = orNode;
+	}
+	FilterTree_Free(filter->filterTree);
+	filter->filterTree = root;
+}
+
+static void _prepareFilterOp(OpFilter *filter) {
+// Filter is applicable, normalize it.
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: Indentation

File: src/execution_plan/optimizations/utilize_indices.c [hunk 8]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
...
+	_normalize_filter(filter->filterTree);
+	// See if the filter tree needed to be modified, if so, replace the original, since the op will be free.
+	if(filter->filterTree->t == FT_N_EXP &&
+	   strcasecmp(filter->filterTree->exp.exp->op.func_name, "in") == 0)
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: Create a local function which performs this check.

File: src/execution_plan/optimizations/utilize_indices.c [hunk 9]
```diff
@@ -195,6 +209,55 @@ bool _simple_predicates(const FT_FilterNode *filter) {
...
+		SIValue v = SI_CloneValue(SIArray_Get(listValue, i));
+		// Ignore everything other than number and strings.
+		if(!(SI_TYPE(v) & (SI_NUMERIC | T_STRING | T_BOOL))) return false;
+	}
+	return true;
+}
+
+static void _transformInToOrSequence(OpFilter *filter) {
+	AR_ExpNode *inOp = filter->filterTree->exp.exp;
+	AR_ExpNode *lhs = rm_malloc(sizeof(AR_ExpNode));
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c]: Please use `AR_EXP_NewOpNode`, `AR_EXP_NewDistinctOpNode`, `AR_EXP_NewVariableOperandNode`
and `AR_EXP_NewConstOperandNode` to construct the new tree.
avoid `rm_malloc` and `memcpy`

File: src/filter_tree/filter_tree.c (hunk 1 above)
Comment (DvirDukhan) [reply src/filter_tree/filter_tree.c]: done

File: src/execution_plan/optimizations/utilize_indices.c (hunk 2 above)
Comment (DvirDukhan) [reply src/execution_plan/optimizations/utilize_indices.c]: removed

File: src/execution_plan/optimizations/utilize_indices.c (hunk 3 above)
Comment (DvirDukhan) [reply src/execution_plan/optimizations/utilize_indices.c]: done

File: src/execution_plan/optimizations/utilize_indices.c (hunk 7 above)
Comment (DvirDukhan) [reply src/execution_plan/optimizations/utilize_indices.c]: done

File: src/execution_plan/optimizations/utilize_indices.c (hunk 8 above)
Comment (DvirDukhan) [reply src/execution_plan/optimizations/utilize_indices.c]: done

File: src/execution_plan/optimizations/utilize_indices.c [hunk 10]
```diff
@@ -267,13 +345,9 @@ OpFilter **_applicableFilters(NodeByLabelScan *scanOp, Index *idx) {
 	while(current->type == OPType_FILTER) {
 		OpFilter *filter = (OpFilter *)current;
 
-		if(_applicableFilter(idx, filter)) {
+		if(_applicableFilter(idx, &filter->filterTree)) {
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/execution_plan/optimizations/utilize_indices.c]: Given the logic in `_simple_predicates`, I think you can just augment that function with:
`if(root->t == FT_N_EXP) return false;`

The subsequent introspection is already sufficient!

**Pull request merged**